src/ucs/debug/debug.c

```diff
@@ -302,42 +331,62 @@ static int get_line_info(struct backtrace_file *file, int backoff,
 
 /**
  * Create a backtrace from the calling location.
- */
-static void ucs_debug_backtrace_create(struct backtrace *bckt)
+ *
+ * @param bckt          Backtrace object.
+ * @param strip         How many frames to strip.
+*/
+void ucs_debug_backtrace_create(backtrace_h *bckt, int strip)
 {
+    size_t size = sizeof(**bckt);
     struct backtrace_file file;
     void *addresses[BACKTRACE_MAX];
     int i, num_addresses;
+    ucs_status_t status;
+
+    *bckt  = NULL;
+    status = ucs_mmap_alloc(&size, (void**)bckt, 0
+                            UCS_MEMTRACK_NAME("debug backtrace object"));
+    if (status != UCS_OK) {
+        ucs_assert(*bckt == NULL);
+        return;
+    }
 
     num_addresses = backtrace(addresses, BACKTRACE_MAX);
 
-    bckt->size     = 0;
-    bckt->position = 0;
+    (*bckt)->size       = 0;
+    (*bckt)->position   = strip;
+    (*bckt)->exclude    = 1;
+    (*bckt)->alloc_size = size;
     for (i = 0; i < num_addresses; ++i) {
         file.dl.address = (unsigned long)addresses[i];
         if (dl_lookup_address(&file.dl) && load_file(&file)) {
-            bckt->size += get_line_info(&file, 1, bckt->lines + bckt->size,
-                                        BACKTRACE_MAX - bckt->size);
+            (*bckt)->size += get_line_info(&file, 1,
+                                           (*bckt)->lines + (*bckt)->size,
+                                           BACKTRACE_MAX - (*bckt)->size);
             unload_file(&file);
         }
     }
-
 }
 
 /**
  * Destroy a backtrace and free all memory.
  *
  * @param bckt          Backtrace object.
  */
-static void ucs_debug_backtrace_destroy(backtrace_h bckt)
+void ucs_debug_backtrace_destroy(backtrace_h bckt)
 {
     int i;
 
+    if (bckt == NULL) {
```

Comment:
why would it be NULL?

Reply:
in case if no memory happened during backtrace creation(valgrind may force no memory) - then backtrace object is NULL and functions backtrace_next & backtrace_destroy just ignore such calls

Reply:
how could valgrind break mmap?